### PR TITLE
KAFKA-16082: Avoid resuming future replica if current replica is in the same directory

### PR DESCRIPTION
**Pull request by gaurav-narula**
KAFKA-16082 describes failure scenarios around replica directory being altered and a log directory/broker crashing.

It is observed that for scenario (3), i.e. a broker crashes while it waits for the future replica to catch up for the second time and the `dir1` is unavailable when the broker is restarted, the broker tries to create the partition in `dir2` according to the metadata in the controller. However, ReplicaManager also tries to resume the stale future replica which was abandoned when the broker crashed. This results in the renaming of the future replica to fail eventually because the directory for the topic partition already exists in `dir2` and the broker then marks `dir2` as offline.

This PR attempts to fix this behaviour by ignoring any future replicas which are in the same directory as where the log exists. It further marks the stale future replica for deletion.

Edit: Instead of removing the abandoned future replica, I've adopted Proven's suggestion to rename the future replica instead and avoid the creation of a new replica in `dir2`.

### CODE REVIEW

**Comment by gaurav-narula** [comment]
CC: @OmniaGM @pprovenzano

**Review by OmniaGM — APPROVED** [review]
This looks very straight forward assuming all tests are passing. LGTM

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 1]
```diff
@@ -1156,6 +1156,32 @@ class LogManager(logDirs: Seq[File],
...
 
+  def maybeRecoverAbandonedFutureLog(topicPartition: TopicPartition, partitionAssignedDirectoryId: Option[Uuid], callback: () => Unit): Unit = {
+    logCreationOrDeletionLock synchronized {
+      for {
+        futureLog <- getLog(topicPartition, isFuture = true)
+        futureDirId <- directoryId(futureLog.parentDir)
+        assignedDirId <- partitionAssignedDirectoryId
+        if futureDirId == assignedDirId
+      } {
+        val sourceLog = futureLog
```
Comment (OmniaGM) [inline core/src/main/scala/kafka/log/LogManager.scala]: `sourceLog` can be moved into the `for` block.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 2]
```diff
@@ -2773,6 +2776,12 @@ class ReplicaManager(val config: KafkaConfig,
         Some(partition, false)
 
       case HostedPartition.None =>
+        var isNew = true
```
Comment (OmniaGM) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: `isNew` can be a val if you changed the code to the following
```
val isNew = if (isLocalFollower) {
           logManager.maybeRecoverAbandonedFutureLog(tp, partitionAssignedDirectoryId, () => {
             isNew = false
           })
         } else true
```

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 3]
```diff
@@ -4831,15 +4831,15 @@ class ReplicaManagerTest {
     try {
       val foo0 = new TopicPartition("foo", 0)
       val emptyDelta = new TopicsDelta(TopicsImage.EMPTY)
-      val (fooPart, fooNew) = replicaManager.getOrCreatePartition(foo0, emptyDelta, FOO_UUID).get
+      val (fooPart, fooNew) = replicaManager.getOrCreatePartition(foo0, emptyDelta, FOO_UUID, false, None).get
```
Comment (OmniaGM) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Any reason why we are not setting `partitionAssignedDirectoryId` to None as a default in `replicaManager.getOrCreatePartition` signature?

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 4]
```diff
@@ -5592,6 +5592,63 @@ class ReplicaManagerTest {
...
+    )
+
+    val directoryId1 = Uuid.randomUuid()
+    val directoryId2 = Uuid.randomUuid()
+    val mockFutureLog = setupMockLog("/KAFKA-16082-test", isFuture = true)
+
+    val mockLogMgr = replicaManager.logManager
+    when(
+      mockLogMgr.getLog(topicPartition, isFuture = true)
+    ).thenReturn {
```
Comment (OmniaGM) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: `{` is used to hold a block but in this case it is a simple value. So maybe replace all `.thenReturn {Some(mockFutureLog)}` to `.thenReturn(Some(mockFutureLog))`

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 5]
```diff
@@ -5592,6 +5592,63 @@ class ReplicaManagerTest {
...
+      Some(mockFutureLog)
+    }
+    when(mockLogMgr.getLog(topicPartition)).thenReturn {
+      None
```
Comment (OmniaGM) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Same `.thenReturn {None}` to `.thenReturn(None)`

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 6]
```diff
@@ -5592,6 +5592,63 @@ class ReplicaManagerTest {
...
+    }
+
+    when(
+      mockLogMgr.directoryId(mockFutureLog.parentDir)
+    ).thenReturn {
+      Some(directoryId1)
```
Comment (OmniaGM) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: same as above

**Review by OmniaGM — COMMENTED** [review]
left few scala nits

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 2 above)
Comment (gaurav-narula) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: I don't think that would work because `isNew` is mutated inside the callback which may be executed conditionally.

Edit: Perhaps I can have `maybeRecoverAbandonedFutureLog` return a boolean instead

**Review by OmniaGM — APPROVED** [review]
LGTM, @pprovenzano @rondagostino can you please have a look?

**Review by pprovenzano — APPROVED** [review]
LGTM

**Comment by OmniaGM** [comment]
Hi, @rondagostino, @showuon and @cmccabe can one of you have a look into this pr please and merge it if you are happy with the changes?

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 7]
```diff
@@ -2859,10 +2867,10 @@ class ReplicaManager(val config: KafkaConfig,
       "local leaders.")
     replicaFetcherManager.removeFetcherForPartitions(localLeaders.keySet)
     localLeaders.forKeyValue { (tp, info) =>
-      getOrCreatePartition(tp, delta, info.topicId).foreach { case (partition, isNew) =>
+      val partitionAssignedDirectoryId = directoryIds.find(_._1.topicPartition() == tp).map(_._2)
+      getOrCreatePartition(tp, delta, info.topicId, isLocalFollower = false, partitionAssignedDirectoryId).foreach { case (partition, isNew) =>
```
Comment (soarez) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: > `isLocalFollower = false`

I don't think this is ok. Setting this to false disables the fix when the broker is the leader. Which can happen in RF=1 or with RF=N when the broker comes back in the ISR (e.g. because there were no new messages, and the other replicas are all shutdown).

I had a conversation with @gaurav-narula about this, and the plan is to apply this fix earlier when the logs are being loaded.

**Review by soarez — CHANGES_REQUESTED** [review]

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 8]
```diff
@@ -2790,7 +2791,6 @@ class ReplicaManager(val config: KafkaConfig,
           //   is unavailable. This is required to ensure that we include the partition's
           //   high watermark in the checkpoint file (see KAFKA-1647).
           val state = info.partition.toLeaderAndIsrPartitionState(tp, isNew)
-          val partitionAssignedDirectoryId = directoryIds.find(_._1.topicPartition() == tp).map(_._2)
```
Comment (soarez) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: It seems the changes in this file are no longer necessary?

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 9]
```diff
@@ -1173,6 +1173,35 @@ class LogManager(logDirs: Seq[File],
...
+  private def findAbandonedFutureLogs(brokerId: Int, newTopicsImage: TopicsImage): Iterable[UnifiedLog] = {
+    futureLogs.values.flatMap { log =>
+      val topicId = log.topicId.getOrElse {
+        throw new RuntimeException(s"The log dir $log does not have a topic ID, " +
+          "which is not allowed when running in KRaft mode.")
+      }
+      val partitionId = log.topicPartition.partition()
+      Option(newTopicsImage.getPartition(topicId, partitionId))
+        .filter(pr => directoryId(log.parentDir).contains(pr.directory(brokerId)))
+        .map(_ => log)
```
Comment (soarez) [inline core/src/main/scala/kafka/log/LogManager.scala]: 1. Do we have the guarantee that the topic will be in the new topics image? Can't the topic be in later metadata delta if a compaction hasn't yet occurred?
2. Shouldn't we check if the main replica is on one of the other directories? If the main replica is on an offline dir (as in the scenario described in the issue) the broker will refuse to start once it is restarted with that directory online, as it will see the two main replicas. If the log directory for the main replica is online, we should be able to detect that here.

**Review by soarez — COMMENTED** [review]
Thanks for the changes. I still think there are some issues.

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 9 above)
Comment (showuon) [reply core/src/main/scala/kafka/log/LogManager.scala]: For (1), I think we'll catch up until the high watermark before publish the image. So it should be safe.
For (2), good suggestion.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 10]
```diff
@@ -2745,10 +2745,10 @@ class ReplicaManager(val config: KafkaConfig,
       "local leaders.")
     replicaFetcherManager.removeFetcherForPartitions(localLeaders.keySet)
     localLeaders.forKeyValue { (tp, info) =>
+      val partitionAssignedDirectoryId = directoryIds.find(_._1.topicPartition() == tp).map(_._2)
       getOrCreatePartition(tp, delta, info.topicId).foreach { case (partition, isNew) =>
         try {
           val state = info.partition.toLeaderAndIsrPartitionState(tp, isNew)
```
Comment (showuon) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: Same here, unnecessary change.

**Review by showuon — COMMENTED** [review]
Thanks for the PR. Left some comments.

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 9 above)
Comment (gaurav-narula) [reply core/src/main/scala/kafka/log/LogManager.scala]: Thanks for the feedback.

For (2), we've couple of options. We can either:

(a) ignore the future replica (say in dir2) if the main replica exists in an online log dir (say dir1) or,
(b) promote the future replica (in dir2)  and remove the main replica (in dir1).

(a) would result in ReplicaManager spawning a replicaAlterLogDir thread for the future replica and correcting the assignment to dir1, only for it to be changed back again to dir2 when the replicaAlterLogDir thread finishes its job. Refer https://github.com/apache/kafka/blob/acecd370cc3b25f12926e7a4664a2648f08c6c9a/core/src/main/scala/kafka/server/ReplicaManager.scala#L2734 and https://github.com/apache/kafka/blob/acecd370cc3b25f12926e7a4664a2648f08c6c9a/core/src/main/scala/kafka/server/ReplicaManager.scala#L2745

Since in these scenarios, the future replica is almost caught up with the main replica, I'm leaning towards option (b) to avoid more reassignments. Please let me know if you feel otherwise.

Comment (soarez) [reply core/src/main/scala/kafka/log/LogManager.scala]: If we didn't know if the future log was caught up or not, then I'd prefer (a), but at this point I can't conceive of a different scenario – other than a failure during replica promotion – that would cause the future log to be in the directory assigned in the metadata. So I agree that the two logs likely will be either caught up or very close. So I agree it makes more sense to do (b) - promote the future log and delete the main one.

We can still run into trouble if the directory with the main replica is offline. At some point that will cause a crash if the directory ever comes back online. But there's nothing we can do about that here. Maybe future work could improve how the broker handles loading conflicting logs.

Comment (gaurav-narula) [reply core/src/main/scala/kafka/log/LogManager.scala]: Addressed in 430546fa5a78118497ada6373607f6a25c78a8e8. Please take a look.

File: core/src/test/java/kafka/testkit/BrokerNode.java [hunk 11]
```diff
@@ -81,8 +82,7 @@ public BrokerNode build(
                     logDataDirectories = Collections.
                         singletonList(String.format("combined_%d", id));
                 } else {
-                    logDataDirectories = Collections.
-                        singletonList(String.format("broker_%d_data0", id));
+                    logDataDirectories = Collections.unmodifiableList(Arrays.asList(String.format("broker_%d_data0", id), String.format("broker_%d_data1", id)));
```
Comment (soarez) [inline core/src/test/java/kafka/testkit/BrokerNode.java]: I'm wondering if this is going to break test cases that don't support JBOD, if so there should be some new failing tests.

File: core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala [hunk 12]
```diff
@@ -1437,7 +1438,9 @@ class KRaftClusterTest {
           assertEquals(Set(1, 2), info.get.isr().asScala.toSet)
         }
 
-        // Modify foo-0 so that it refers to a future replica
+        // Modify foo-0 so that it refers to a future replica.
+        // This has the same effect as the main replica being in an offline
+        // log dir and the broker crashing just at the time of promotion
```
Comment (soarez) [inline core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala]: I think I know what you mean, but I don't know if this is easy to understand in its current phrasing. Here's a suggestion:

This is equivalent to a failure during the promotion of the future replica and a restart with directory for the main replica being offline.

File: core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala [hunk 13]
```diff
@@ -1456,6 +1459,76 @@ class KRaftClusterTest {
...
+        val log = broker0.logManager.getLog(foo0).get
+
+        // Copy foo-0 to another log dir
+        val parentDir = log.parentDir
+        var targetParentDir = parentDir.substring(0, parentDir.length - 1)
+        if (parentDir.endsWith("0")) {
+          targetParentDir += "1"
+        } else {
+          targetParentDir += "0"
+        }
```
Comment (soarez) [inline core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala]: This assumes your change to `BrokerNode.build()` and its specific naming pattern. We'd get a more robust test if instead you access the directory list on the broker, filter out `parentDir`, and choose one of the remaining log dirs.

File: core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala [hunk 14]
```diff
@@ -1456,6 +1459,76 @@ class KRaftClusterTest {
...
+        broker0.shutdown()
+        TestUtils.retry(60000) {
+          val info = broker1.metadataCache.getPartitionInfo("foo", 0)
+          assertTrue(info.isDefined)
+          assertEquals(Set(1, 2), info.get.isr().asScala.toSet)
+        }
+
+        val log = broker0.logManager.getLog(foo0).get
+
+        // Copy foo-0 to another log dir
```
Comment (soarez) [inline core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala]: It could be helpful for future eyes on this, to leave a comment explaining why we're manually copying the log dir in the test.

**Review by soarez — APPROVED** [review]
Thanks for the changes. The test failures seem to be unrelated. 
LGTM

**Comment by soarez** [comment]
@showuon PTAL

**Comment by showuon** [comment]
Will check it again within this week. Thanks.

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 15]
```diff
@@ -1176,6 +1176,42 @@ class LogManager(logDirs: Seq[File],
...
+        log.removeLogMetrics()
+        log.renameDir(UnifiedLog.logDeleteDirName(tp), shouldReinitialize = false)
+        addLogToBeDeleted(log)
+        info(s"Old log for partition ${tp} is renamed to ${log.dir.getAbsolutePath} and is scheduled for deletion")
+      }
+
+      currentLogs.put(tp, futureLog)
+      futureLog.newMetrics()
+
+      info(s"Successfully renamed abandoned future log for $tp")
```
Comment (showuon) [inline core/src/main/scala/kafka/log/LogManager.scala]: I can see we basically do the similar things as what `replaceCurrentWithFutureLog` did, why should we duplicate the codes here again? Could we not be able to call `replaceCurrentWithFutureLog` with additional parameters to indicate we want to skip some actions?

Like what we did now, we didn't close the `currentLog`, which should have potential resource leak. Using the `replaceCurrentWithFutureLog` could avoid this issue happening. WDYT?

**Review by showuon — COMMENTED** [review]
@gaurav-narula , thanks for the PR. Left one comment.

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 9 above)
Comment (showuon) [reply core/src/main/scala/kafka/log/LogManager.scala]: I agree option (b) is better. But for this:
> We can still run into trouble if the directory with the main replica is offline. At some point that will cause a crash if the directory ever comes back online. But there's nothing we can do about that here. Maybe future work could improve how the broker handles loading conflicting logs.

I believe this PR: https://github.com/apache/kafka/pull/15335 should already fix this issue. @OmniaGM , could you help confirm it? If not, I think we should not promote the future log when main replica is offline to cause potential future issue.

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 15 above)
Comment (chia7712) [reply core/src/main/scala/kafka/log/LogManager.scala]: it seems to me that we need to call `closeHandlers` at least if we don't do refactor in this PR.

File: core/src/test/java/kafka/testkit/BrokerNode.java (hunk 11 above)
Comment (chia7712) [reply core/src/test/java/kafka/testkit/BrokerNode.java]: Personally, I'd like to change setNumBrokerNodes(int) to setBrokerNodes(int, int). The second argument is used to define the number of data folders. After all, not all tests require 2+ folders. Maybe we can address that in another PR.

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 16]
```diff
@@ -1176,6 +1176,42 @@ class LogManager(logDirs: Seq[File],
...
+
+      futureLog.renameDir(UnifiedLog.logDirName(tp), shouldReinitialize = true)
+      futureLog.removeLogMetrics()
+      futureLogs.remove(tp)
+
+      currentLog.foreach { log =>
+        log.removeLogMetrics()
+        log.renameDir(UnifiedLog.logDeleteDirName(tp), shouldReinitialize = false)
+        addLogToBeDeleted(log)
+        info(s"Old log for partition ${tp} is renamed to ${log.dir.getAbsolutePath} and is scheduled for deletion")
```
Comment (chia7712) [inline core/src/main/scala/kafka/log/LogManager.scala]: `${tp}` -> `$tp`

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 15 above)
Comment (gaurav-narula) [reply core/src/main/scala/kafka/log/LogManager.scala]: That's a great suggestion! I'm going to try refactoring this to use `replaceCurrentWithFutureLog`

File: core/src/test/java/kafka/testkit/BrokerNode.java (hunk 11 above)
Comment (chia7712) [reply core/src/test/java/kafka/testkit/BrokerNode.java]: I have created https://issues.apache.org/jira/browse/KAFKA-16559

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 15 above)
Comment (gaurav-narula) [reply core/src/main/scala/kafka/log/LogManager.scala]: I've refactored LogManager to use `replaceCurrentWithFutureLog` in [b87a21f](https://github.com/apache/kafka/pull/15136/commits/b87a21f3bd0eea6e4083a2d14b41053361f7b40a). PTAL

File: core/src/test/java/kafka/testkit/BrokerNode.java (hunk 11 above)
Comment (gaurav-narula) [reply core/src/test/java/kafka/testkit/BrokerNode.java]: Rebased to use the new method in the builder

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 17]
```diff
@@ -1189,50 +1216,61 @@ class LogManager(logDirs: Seq[File],
...
+    }
 
-      try {
-        sourceLog.renameDir(UnifiedLog.logDeleteDirName(topicPartition), shouldReinitialize = true)
+    // Now that future replica has been successfully renamed to be the current replica
+    // Update the cached map and log cleaner as appropriate.
+    futureLogs.remove(topicPartition)
+    currentLogs.put(topicPartition, destLog)
+    if (cleaner != null) {
+      cleaner.alterCheckpointDir(topicPartition, sourceLog.map(_.parentDirFile), destLog.parentDirFile)
```
Comment (chia7712) [inline core/src/main/scala/kafka/log/LogManager.scala]: It seems `cleaner.alterCheckpointDir` will do nothing if `sourceLog` is empty. Maybe we can revert those changes and run `alterCheckpointDir` only if `sourceLog` is defined.

**Review by chia7712 — COMMENTED** [review]
@gaurav-narula thanks for refactoring code. one major comment left. please take a look.

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 17 above)
Comment (gaurav-narula) [reply core/src/main/scala/kafka/log/LogManager.scala]: Addressed in [d02fc0b](https://github.com/apache/kafka/pull/15136/commits/d02fc0b9abeacf95096d4275232ea0cf829d836a)

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 18]
```diff
@@ -1189,50 +1216,63 @@ class LogManager(logDirs: Seq[File],
...
-      // Now that future replica has been successfully renamed to be the current replica
-      // Update the cached map and log cleaner as appropriate.
-      futureLogs.remove(topicPartition)
-      currentLogs.put(topicPartition, destLog)
-      if (cleaner != null) {
-        cleaner.alterCheckpointDir(topicPartition, sourceLog.parentDirFile, destLog.parentDirFile)
-        resumeCleaning(topicPartition)
+  def replaceCurrentWithFutureLog(sourceLog: Option[UnifiedLog], destLog: UnifiedLog, updateHighWatermark: Boolean = false): Unit = {
+    val topicPartition = destLog.topicPartition
+    info(s"Attempting to replace current log $sourceLog with $destLog for $topicPartition")
```
Comment (chia7712) [inline core/src/main/scala/kafka/log/LogManager.scala]: `sourceLog` could be empty now, so maybe we need to revise the log message

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 19]
```diff
@@ -1178,6 +1178,33 @@ class LogManager(logDirs: Seq[File],
     }
   }
 
+  def recoverAbandonedFutureLogs(brokerId: Int, newTopicsImage: TopicsImage): Unit = {
+    val abandonedFutureLogs = findAbandonedFutureLogs(brokerId, newTopicsImage)
+    abandonedFutureLogs.foreach { case (futureLog, currentLog) =>
+      val tp = futureLog.topicPartition
+      if (cleaner != null) {
```
Comment (chia7712) [inline core/src/main/scala/kafka/log/LogManager.scala]: we can replace this `abortAndPauseCleaning(tp)`

Comment (chia7712) [reply core/src/main/scala/kafka/log/LogManager.scala]: I guess `cleaner.abortAndPauseCleaning` is added because we call `resumeCleaning` later, and it will cause error if we don't call `abortAndPauseCleaning` here?

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 20]
```diff
@@ -1178,6 +1178,33 @@ class LogManager(logDirs: Seq[File],
...
+        cleaner.abortAndPauseCleaning(tp)
+      }
+
+      replaceCurrentWithFutureLog(currentLog, futureLog)
+
+      info(s"Successfully renamed abandoned future log for $tp")
```
Comment (chia7712) [inline core/src/main/scala/kafka/log/LogManager.scala]: line#1189 will print `info(s"The current replica is successfully replaced with the future replica for $topicPartition")` and that is not much related to fact, right?

**Review by chia7712 — COMMENTED** [review]
@gaurav-narula thanks for updated PR. please take a look at following comments.

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 19 above)
Comment (gaurav-narula) [reply core/src/main/scala/kafka/log/LogManager.scala]: > I guess cleaner.abortAndPauseCleaning is added because we call resumeCleaning later, and it will cause error if we don't call abortAndPauseCleaning here?

That's correct if the cleaning operation hasn't started. The cleaning operation is scheduled on a separate thread so we cannot be sure if `inProgress` map in `LogCleanerManager` has a key for the given topicPartition at the time we iterate over these logs.

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 18 above)
Comment (gaurav-narula) [reply core/src/main/scala/kafka/log/LogManager.scala]: Addressed this and other comments in [31412a9](https://github.com/apache/kafka/pull/15136/commits/31412a9485b63731a49b5d44d6dc6fbeaf52dd0f)

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 19 above)
Comment (chia7712) [reply core/src/main/scala/kafka/log/LogManager.scala]: Maybe we need to add comments for `abortAndPauseCleaning`. The `recoverAbandonedFutureLogs` is called in starting and calling `abortAndPauseCleaning` here is a bit weird to me

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 21]
```diff
@@ -1189,49 +1217,62 @@ class LogManager(logDirs: Seq[File],
...
-      // we will add metrics back after sourceLog remove the metrics
-      destLog.removeLogMetrics()
-      destLog.updateHighWatermark(sourceLog.highWatermark)
+      info(s"Attempting to replace current log $sourceLog with $destLog for $topicPartition")
+      replaceCurrentWithFutureLog(Option(sourceLog), destLog, updateHighWatermark = true)
+      info(s"The current replica is successfully replaced with the future replica for $topicPartition")
+    }
+  }
+
+  def replaceCurrentWithFutureLog(sourceLog: Option[UnifiedLog], destLog: UnifiedLog, updateHighWatermark: Boolean = false): Unit = {
```
Comment (chia7712) [inline core/src/main/scala/kafka/log/LogManager.scala:1233]: it seems we don't need `updateHighWatermark`, since we call `updateHighWatermark` only if `sourceLog` is defined.

**Review by chia7712 — COMMENTED** [review]
@gaurav-narula please take a look at two minor comments. thanks!

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 21 above)
Comment (gaurav-narula) [reply core/src/main/scala/kafka/log/LogManager.scala:1233]: I think that's not necessarily true. `findAbandonedFutureLogs` may find a `sourceLog` if the log directory is online but I don't think it's safe to update the high watermark even then because sourceLog may be ahead of futureLog.

Comment (chia7712) [reply core/src/main/scala/kafka/log/LogManager.scala:1233]: > I think that's not necessarily true. findAbandonedFutureLogs may find a sourceLog if the log directory is online but I don't think it's safe to update the high watermark even then because sourceLog may be ahead of futureLog.

you are right!

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 19 above)
Comment (gaurav-narula) [reply core/src/main/scala/kafka/log/LogManager.scala]: Added a comment in [3d78e55](https://github.com/apache/kafka/pull/15136/commits/3d78e5567fd16aa77f0b9f8c510ad26f0e6443c6)

**Review by chia7712 — APPROVED** [review]
LGTM

**Comment by chia7712** [comment]
the failed `testParseAndValidateAddressesWithReverseLookup` is traced by #15758. will merge it
